cmd/libsnap-confine-private/infofile.c

```diff
@@ -0,0 +1,215 @@
+        err = sc_error_init(SC_LIBSNAP_ERROR, SC_BUG, "caller_state cannot be NULL");
+        goto out;
+    }
+
+    if (sc_streq(caller_state->wanted_key, scanner_state->key)) {
+        caller_state->stored_value = sc_strdup(scanner_state->value);
+        scanner_state->stop = true;
+    }
+
+out:
+    return sc_error_forward(err_out, err);
+}
+
+int sc_infofile_get_key(FILE *stream, const char *key, char **value, sc_error **err_out) {
+    sc_error *err = NULL;
+
+    /* NOTE: stream is checked by sc_infofile_scan */
+    if (key == NULL) {
+        err = sc_error_init(SC_LIBSNAP_ERROR, SC_API_MISUSE, "key cannot be NULL");
+        goto out;
+    }
+    if (value == NULL) {
+        err = sc_error_init(SC_LIBSNAP_ERROR, SC_API_MISUSE, "value cannot be NULL");
+        goto out;
+    }
+
+    sc_infofile_get_key_state scanner_state = {.wanted_key = key};
+    sc_infofile_scanner_fn scanner_fn = sc_infofile_get_key_scanner;
+    /* Call the scanner and ignore the error code. We will forward any
+     * error in a moment, but first we want to store the result of the
+     * search, successful or not, so that *value is always defined. */
+    (void)sc_infofile_scan(stream, scanner_fn, &scanner_state, &err);
+    *value = scanner_state.stored_value;
+
+out:
+    return sc_error_forward(err_out, err);
+}
+
+static int sc_infofile_scan(FILE *stream, sc_infofile_scanner_fn scanner_fn, void *caller_state, sc_error **err_out) {
+    sc_error *err = NULL;
+    size_t line_size = 0;
+    char *line_buf SC_CLEANUP(sc_cleanup_string) = NULL;
+
+    if (stream == NULL) {
+        err = sc_error_init(SC_LIBSNAP_ERROR, SC_API_MISUSE, "stream cannot be NULL");
+        goto out;
+    }
+    if (scanner_fn == NULL) {
+        err = sc_error_init(SC_LIBSNAP_ERROR, SC_API_MISUSE, "scanner_fn cannot be NULL");
+        goto out;
+    }
+
+    /* This loop advances through subsequent lines. */
+    for (int lineno = 1;; ++lineno) {
+        errno = 0;
+        ssize_t nread = getline(&line_buf, &line_size, stream);
+        if (nread < 0 && errno != 0) {
+            err = sc_error_init_from_errno(errno, "cannot read beyond line %d", lineno);
+            goto out;
+        }
+        if (nread <= 0) {
+            break; /* There is nothing more to read. */
+        }
+        /* NOTE: beyond this line the buffer is never empty. */
+
+        /* Guard against malformed input that may contain NUL bytes that
+         * would confuse the code below. */
+        if (memchr(line_buf, '\0', nread) != NULL) {
+            err = sc_error_init(SC_LIBSNAP_ERROR, 0, "line %d contains NUL byte", lineno);
+            goto out;
+        }
+        /* Replace the newline character, if any, with the NUL byte. */
+        if (line_buf[nread - 1] == '\n') {
+            line_buf[nread - 1] = '\0';
+        }
+        /* Guard against malformed input that does not contain '=' byte */
+        char *eq_ptr = memchr(line_buf, '=', nread);
+        if (eq_ptr == NULL) {
```

Comment:
This could probably be: `if (eq_ptr == NULL || eq_ptr == line_buf)`

Reply:
Oh, interesting. I'll add this case and test for it.

Reply:
Done